### ISSUE TITLE
[YSQL] Very high query latency with default settings in a variety of configs

### ISSUE THREAD

**Issue by jfrconley** [yugabyte/yugabyte-db#10654]
Jira Link: [DB-1295](https://yugabyte.atlassian.net/browse/DB-1295)
### Description

I've been trying to evaluate YugabyteDB for use as our primary database, but I'm running into some issues. For some reason I'm getting extremely high (100-200ms) read latencies for simple select queries on fairly small amounts of data (several thousand records).

I have tried several different configurations:
EKS cluster using the helm config and 3 m5ad large instances
EKS cluster using the helm config and 3 m5ad xlarge instances
Simple docker setup using yugabyted separately on 2 different machines

My data source is several tables using shared compound primary keys.

My test query is a simple join between two tables
```sql
select id, supports
from firmware_support,
     device
where device.tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'
  and firmware_support.tenant_id = device.tenant_id
  and firmware_support.device_id = device.id
order by device.tenant_id, device.id
limit 100;
```

This generates the following plan
```
QUERY PLAN
Limit  (cost=0.00..26.90 rows=50 width=322) (actual time=4.356..112.669 rows=100 loops=1)
  ->  Nested Loop  (cost=0.00..26.90 rows=50 width=322) (actual time=4.354..112.562 rows=100 loops=1)
        ->  Index Only Scan using device_keyset on device  (cost=0.00..15.25 rows=100 width=290) (actual time=2.959..3.531 rows=100 loops=1)
              Index Cond: (tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid)
              Heap Fetches: 0
        ->  Index Scan using firmware_support_pk on firmware_support  (cost=0.00..0.12 rows=1 width=322) (actual time=1.030..1.030 rows=1 loops=100)
              Index Cond: ((tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid) AND ((device_id)::text = (device.id)::text))
Planning Time: 0.416 ms
Execution Time: 113.672 ms
```

It is using index only scans over 100 rows and is somehow taking an absurd 113ms. I noticed that the Nested Loop takes up the majority of the time. This specific plan was generated on a local docker container with 8 cpu cores, 8gb of ram, and an NVME ssd but almost identical results were provided by the larger clusters as well. I can only assume that something is misconfigured here as these results do not line up with what I would expect given the benchmarks I've seen. CockroachDB was able to complete this same query in 8ms, which is not amazing but is fairly acceptable.

I've been trying to get this working for several days now, and I can't seem to figure out what I'm doing wrong. Is there some setting or configuration that I'm missing? I would have expected out of the box performance to be much better.

EDIT:
Using `set enable_nestloop=false` brought the execution time down to 20ms. A significant improvement, but still not great.
```sql
QUERY PLAN
Limit  (cost=18.57..35.20 rows=100 width=596) (actual time=22.155..22.245 rows=100 loops=1)
  ->  Merge Left Join  (cost=18.57..35.20 rows=100 width=596) (actual time=22.153..22.234 rows=100 loops=1)
        Merge Cond: ((device.id)::text = (fs.device_id)::text)
        Join Filter: (device.tenant_id = fs.tenant_id)
        ->  Index Only Scan using device_keyset on device  (cost=0.00..15.25 rows=100 width=290) (actual time=2.776..2.813 rows=100 loops=1)
              Index Cond: (tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid)
              Heap Fetches: 0
        ->  Sort  (cost=18.57..18.82 rows=100 width=322) (actual time=19.369..19.373 rows=100 loops=1)
              Sort Key: fs.device_id
              Sort Method: quicksort  Memory: 328kB
              ->  Index Scan using firmware_support_pk on firmware_support fs  (cost=0.00..15.25 rows=100 width=322) (actual time=11.551..18.928 rows=1988 loops=1)
                    Index Cond: (tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid)
Planning Time: 0.384 ms
Execution Time: 22.443 ms

```

EDIT 2:
Original query plan with full output. Indeed, it does seem like the Nested loop join is taking up over a 100ms.
```json
[
  {
    "Plan": {
      "Node Type": "Limit",
      "Parallel Aware": false,
      "Startup Cost": 0.00,
      "Total Cost": 26.90,
      "Plan Rows": 50,
      "Plan Width": 322,
      "Actual Startup Time": 4.087,
      "Actual Total Time": 102.344,
      "Actual Rows": 100,
      "Actual Loops": 1,
      "Shared Hit Blocks": 0,
      "Shared Read Blocks": 0,
      "Shared Dirtied Blocks": 0,
      "Shared Written Blocks": 0,
      "Local Hit Blocks": 0,
      "Local Read Blocks": 0,
      "Local Dirtied Blocks": 0,
      "Local Written Blocks": 0,
      "Temp Read Blocks": 0,
      "Temp Written Blocks": 0,
      "Plans": [
        {
          "Node Type": "Nested Loop",
          "Parent Relationship": "Outer",
          "Parallel Aware": false,
          "Join Type": "Inner",
          "Startup Cost": 0.00,
          "Total Cost": 26.90,
          "Plan Rows": 50,
          "Plan Width": 322,
          "Actual Startup Time": 4.084,
          "Actual Total Time": 102.247,
          "Actual Rows": 100,
          "Actual Loops": 1,
          "Inner Unique": true,
          "Shared Hit Blocks": 0,
          "Shared Read Blocks": 0,
          "Shared Dirtied Blocks": 0,
          "Shared Written Blocks": 0,
          "Local Hit Blocks": 0,
          "Local Read Blocks": 0,
          "Local Dirtied Blocks": 0,
          "Local Written Blocks": 0,
          "Temp Read Blocks": 0,
          "Temp Written Blocks": 0,
          "Plans": [
            {
              "Node Type": "Index Only Scan",
              "Parent Relationship": "Outer",
              "Parallel Aware": false,
              "Scan Direction": "Forward",
              "Index Name": "device_keyset",
              "Relation Name": "device",
              "Alias": "device",
              "Startup Cost": 0.00,
              "Total Cost": 15.25,
              "Plan Rows": 100,
              "Plan Width": 290,
              "Actual Startup Time": 2.796,
              "Actual Total Time": 3.218,
              "Actual Rows": 100,
              "Actual Loops": 1,
              "Index Cond": "(tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid)",
              "Rows Removed by Index Recheck": 0,
              "Heap Fetches": 0,
              "Shared Hit Blocks": 0,
              "Shared Read Blocks": 0,
              "Shared Dirtied Blocks": 0,
              "Shared Written Blocks": 0,
              "Local Hit Blocks": 0,
              "Local Read Blocks": 0,
              "Local Dirtied Blocks": 0,
              "Local Written Blocks": 0,
              "Temp Read Blocks": 0,
              "Temp Written Blocks": 0
            },
            {
              "Node Type": "Index Only Scan",
              "Parent Relationship": "Inner",
              "Parallel Aware": false,
              "Scan Direction": "Forward",
              "Index Name": "firmware_support_tenant_id_device_id_supports_idx",
              "Relation Name": "firmware_support",
              "Alias": "firmware_support",
              "Startup Cost": 0.00,
              "Total Cost": 0.12,
              "Plan Rows": 1,
              "Plan Width": 322,
              "Actual Startup Time": 0.942,
              "Actual Total Time": 0.942,
              "Actual Rows": 1,
              "Actual Loops": 100,
              "Index Cond": "((tenant_id = '3bf50ecf-2f7f-469f-88a4-2aa88bdc083b'::uuid) AND (device_id = (device.id)::text))",
              "Rows Removed by Index Recheck": 0,
              "Heap Fetches": 0,
              "Shared Hit Blocks": 0,
              "Shared Read Blocks": 0,
              "Shared Dirtied Blocks": 0,
              "Shared Written Blocks": 0,
              "Local Hit Blocks": 0,
              "Local Read Blocks": 0,
              "Local Dirtied Blocks": 0,
              "Local Written Blocks": 0,
              "Temp Read Blocks": 0,
              "Temp Written Blocks": 0
            }
          ]
        }
      ]
    },
    "Planning Time": 0.414,
    "Triggers": [
    ],
    "Execution Time": 103.045
  }
]
```

EDIT 3:
Just tried the same queries on a free yugabyte cloud cluster with a single cpu core and execution times were less than half. I'm pretty sure now there's just something I'm misconfiguring and it can't be resource, or network latency related.

**Comment by kmuthukk** [comment]
Addition info shared by @jfrconley :

```
postgres=# \d device
                           Table "public.device"
   Column   |          Type          | Collation | Nullable |   Default    
------------+------------------------+-----------+----------+--------------
 tenant_id  | uuid                   |           | not null | 
 id         | character varying(128) |           | not null | 
 image      | character varying(512) |           |          | 
 sub_type   | character varying(64)  |           | not null | 
 type       | device_types           |           | not null | 
 name       | character varying(128) |           |          | 
 created_at | date                   |           |          | CURRENT_DATE
 updated_at | date                   |           |          | 
Indexes:
    "device_pk" PRIMARY KEY, lsm (tenant_id HASH, id ASC)
    "device_keyset" lsm (tenant_id HASH, id ASC)
Referenced by:
    TABLE "device_data_ble" CONSTRAINT "device_data_ble_tenant_id_fkey" FOREIGN KEY (tenant_id, gateway_id) REFERENCES device(tenant_id, id) ON UPDATE RESTRICT
    TABLE "device_data_ble" CONSTRAINT "device_data_ble_tenant_id_fkey1" FOREIGN KEY (tenant_id, device_id) REFERENCES device(tenant_id, id) ON UPDATE RESTRICT ON DELETE CASCADE
    TABLE "device_version" CONSTRAINT "device_version_tenant_id_fkey" FOREIGN KEY (tenant_id, device_id) REFERENCES device(tenant_id, id) ON DELETE CASCADE
    TABLE "firmware_support" CONSTRAINT "firmware_support_tenant_id_fkey" FOREIGN KEY (tenant_id, device_id) REFERENCES device(tenant_id, id) ON UPDATE RESTRICT ON DELETE CASCADE
    TABLE "tag_association" CONSTRAINT "tag_association_tenant_id_fkey" FOREIGN KEY (tenant_id, device_id) REFERENCES device(tenant_id, id) ON DELETE CASCADE
Triggers:
    device_set_update_version AFTER INSERT OR UPDATE ON device FOR EACH ROW EXECUTE PROCEDURE trigger_device_update_version()
    device_set_updated_time BEFORE UPDATE ON device FOR EACH ROW EXECUTE PROCEDURE trigger_set_updated_time()

postgres=# \d firmware_support
                              Table "public.firmware_support"
    Column     |          Type          | Collation | Nullable |          Default          
---------------+------------------------+-----------+----------+---------------------------
 tenant_id     | uuid                   |           | not null | 
 device_id     | character varying(128) |           | not null | 
 app_name      | character varying(128) |           |          | 
 app_version   | character varying(128) |           |          | 
 ble_support   | boolean                |           |          | false
 boot_version  | character varying(128) |           |          | 
 modem_version | character varying(128) |           |          | 
 supports      | firmware_types[]       |           |          | ARRAY[]::firmware_types[]
 created_at    | date                   |           |          | CURRENT_DATE
 updated_at    | date                   |           |          | 
Indexes:
    "firmware_support_pk" PRIMARY KEY, lsm (tenant_id HASH, device_id ASC)
    "firmware_support_tenant_id_device_id_supports_idx" UNIQUE, lsm (tenant_id HASH, device_id ASC) INCLUDE (supports)
Foreign-key constraints:
    "firmware_support_tenant_id_fkey" FOREIGN KEY (tenant_id, device_id) REFERENCES device(tenant_id, id) ON UPDATE RESTRICT ON DELETE CASCADE
Triggers:
    firmware_support_set_update_version AFTER UPDATE ON firmware_support FOR EACH ROW EXECUTE PROCEDURE trigger_update_only_device_version()
    firmware_support_set_updated_time BEFORE UPDATE ON firmware_support FOR EACH ROW EXECUTE PROCEDURE trigger_set_updated_time()
```

**Comment by kmuthukk** [comment]
The two tables above have 1:1 relation between rows.

Was able to repro issue by creating two tables like this:

```
CREATE TABLE IF NOT EXISTS users(
                      tenant_id     text,
                      user_id       text,
                      age    int,
                      city   text,
                      PRIMARY KEY(tenant_id, user_id))

CREATE TABLE IF NOT EXISTS balance(
                      tenant_id     text,
                      user_id       text,
                      balance    int,
                      PRIMARY KEY(tenant_id, user_id))
```

and insert rows in both tables for 100 tenants and 100 users each. [This is the script](https://gist.github.com/kmuthukk/2c4934c0f43a82297811e79f8d414e4a) used for loading.

Default plan of nested loop is not ideal, but disabling it seems to do the job well:

```
yugabyte=# EXPLAIN ANALYZE
SELECT U.user_id, U.city, B.balance
  FROM users U, balance B
WHERE  U.tenant_id = 'tenant-0-87'
  AND  U.tenant_id = B.tenant_id
  AND  U.user_id = B.user_id;

yugabyte-# yugabyte-# yugabyte-# yugabyte-# yugabyte-#                                                              QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.00..26.90 rows=50 width=68) (actual time=1.892..36.183 rows=100 loops=1)
   ->  Index Scan using balance_pkey on balance b  (cost=0.00..15.25 rows=100 width=68) (actual time=1.271..1.448 rows=100 loops=1)
         Index Cond: (tenant_id = 'tenant-0-87'::text)
   ->  Index Scan using users_pkey on users u  (cost=0.00..0.12 rows=1 width=96) (actual time=0.328..0.328 rows=1 loops=100)
         Index Cond: ((tenant_id = 'tenant-0-87'::text) AND (user_id = b.user_id))
 Planning Time: 22.540 ms
 Execution Time: 36.438 ms
(7 rows)

yugabyte=# set enable_nestloop=false;
SET

yugabyte=# EXPLAIN ANALYZE
SELECT U.user_id, U.city, B.balance
  FROM users U, balance B
WHERE  U.tenant_id = 'tenant-0-87'
  AND  U.tenant_id = B.tenant_id
  AND  U.user_id = B.user_id;

yugabyte-# yugabyte-# yugabyte-# yugabyte-# yugabyte-#                                                              QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------
 Merge Join  (cost=0.00..31.50 rows=50 width=68) (actual time=1.690..2.034 rows=100 loops=1)
   Merge Cond: (u.user_id = b.user_id)
   ->  Index Scan using users_pkey on users u  (cost=0.00..15.25 rows=100 width=96) (actual time=0.977..1.098 rows=100 loops=1)
         Index Cond: (tenant_id = 'tenant-0-87'::text)
   ->  Index Scan using balance_pkey on balance b  (cost=0.00..15.25 rows=100 width=68) (actual time=0.701..0.792 rows=100 loops=1)
         Index Cond: (tenant_id = 'tenant-0-87'::text)
 Planning Time: 0.215 ms
 Execution Time: 2.958 ms
(8 rows)
```

**Comment by kmuthukk** [comment]
I also tried my test case on 2.6.1, and with disabling nestloop. There it uses hash join... (so maybe we enabled merge join more recently).. but timing is pretty good, similar to merge join.

```
yugabyte=# set enable_nestloop=false;
SET
Time: 0.261 ms
yugabyte=# EXPLAIN ANALYZE
SELECT U.user_id, U.city, B.balance
  FROM users U, balance B
WHERE  U.tenant_id = 'tenant-0-87'
  AND  U.tenant_id = B.tenant_id
yugabyte-# yugabyte-# yugabyte-# yugabyte-# yugabyte-#   AND  U.user_id = B.user_id;
                                                                QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=16.50..32.01 rows=50 width=68) (actual time=2.516..2.634 rows=100 loops=1)
   Hash Cond: (u.user_id = b.user_id)
   ->  Index Scan using users_pkey on users u  (cost=0.00..15.25 rows=100 width=96) (actual time=1.166..1.233 rows=100 loops=1)
         Index Cond: (tenant_id = 'tenant-0-87'::text)
   ->  Hash  (cost=15.25..15.25 rows=100 width=68) (actual time=0.816..0.816 rows=100 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 14kB
         ->  Index Scan using balance_pkey on balance b  (cost=0.00..15.25 rows=100 width=68) (actual time=0.712..0.781 rows=100 loops=1)
               Index Cond: (tenant_id = 'tenant-0-87'::text)
 Planning Time: 0.208 ms
 Execution Time: 2.719 ms
(10 rows)

Time: 3.332 ms
```

CC: @andrei-mart , @tanujnay112 , @ramsrivatsa , @m-iancu

**Comment by FranckPachot** [comment]
The problem if disabling Nested Loop the index access cannot use user_id which is ok here but not in the general case.
I've rewritten @kmuthukk query to keep the nested loop, but reading the inner table with an array from the outer table:
```
yugabyte=# EXPLAIN ANALYZE
           with U as (
            -- this is the outer table
            select * from users where tenant_id = 'tenant-0-27'
           ),
           B as (
            -- this is the inner table
            select * from balance
            -- I manually join with an array from the outer table
            where tenant_id=any(array( select tenant_id from U ))
            and   user_id=  any(array( select user_id   from U ))
           )
           select * from U join B using (tenant_id,user_id);
           
                                                           QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=23.42..26.18 rows=1 width=104) (actual time=3.827..3.896 rows=100 loops=1)
   Hash Cond: ((u.tenant_id = b.tenant_id) AND (u.user_id = b.user_id))
   CTE u
     ->  Index Scan using users_pkey on users  (cost=0.00..15.25 rows=100 width=100) (actual time=1.930..2.000 rows=100 loops=1)
           Index Cond: (tenant_id = 'tenant-0-27'::text)
   CTE b
     ->  Index Scan using balance_pkey on balance  (cost=4.00..8.13 rows=1 width=68) (actual time=1.601..1.813 rows=100 loops=1)
           Index Cond: ((tenant_id = ANY ($1)) AND (user_id = ANY ($2)))
           InitPlan 2 (returns $1)
             ->  CTE Scan on u u_1  (cost=0.00..2.00 rows=100 width=32) (actual time=0.001..0.100 rows=100 loops=1)
           InitPlan 3 (returns $2)
             ->  CTE Scan on u u_2  (cost=0.00..2.00 rows=100 width=32) (actual time=0.000..0.013 rows=100 loops=1)
   ->  CTE Scan on u  (cost=0.00..2.00 rows=100 width=100) (actual time=1.932..1.944 rows=100 loops=1)
   ->  Hash  (cost=0.02..0.02 rows=1 width=68) (actual time=1.885..1.885 rows=100 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 14kB
         ->  CTE Scan on b  (cost=0.00..0.02 rows=1 width=68) (actual time=1.603..1.851 rows=100 loops=1)
 Planning Time: 0.187 ms
 Execution Time: 3.973 ms
(18 rows)
```

It could be great if the query planner can do this kind of transformation, but I think in the current version Nested Loop doesn't batch the access to the inner table. A consequence is that it does not preserver the order of the outer table, so there are some cases in Top-N queries where the row-by-row loop may be preferred.

**Comment by andrei-mart** [comment]
> I also tried my test case on 2.6.1, and with disabling nestloop. There it uses hash join... (so maybe we enabled merge join more recently).. but timing is pretty good, similar to merge join.
> 

Correct, merge join support landed in 2.7

https://github.com/yugabyte/yugabyte-db/commit/6b0bcc706f10f16faa6635a6d0b55d17e0f68419

**Comment by timothy-e** [comment]
I was playing around with my query stuff, trying to see it's impact on JOINs. I ended up creating a simpler test case for this issue. 
```sql
CREATE TABLE t_1000 (id int PRIMARY KEY, val int);
INSERT INTO t_1000 SELECT i, i FROM generate_series(1, 1000) AS i;
EXPLAIN ANALYZE SELECT * FROM t_1000 AS a, t_1000 AS b WHERE a.id = b.id;
```
gives the plan
```
 Nested Loop  (cost=0.00..213.89 rows=1000 width=16) (actual time=2.741..369.702 rows=1000 loops=1)
   ->  Seq Scan on t_1000 a  (cost=0.00..100.00 rows=1000 width=8) (actual time=2.313..3.879 rows=1000 loops
=1)
   ->  Index Scan using t_1000_pkey on t_1000 b  (cost=0.00..0.11 rows=1 width=8) (actual time=0.347..0.347
rows=1 loops=1000)
         Index Cond: (id = a.id)
 Planning Time: 2.499 ms
 Execution Time: 370.219 ms
(6 rows)
```
and if we disable nestloop:
```sql
SET enable_nestloop = FALSE;
EXPLAIN ANALYZE SELECT * FROM t_1000 AS a, t_1000 AS b WHERE a.id = b.id;
```
then we get the much faster plan:
```
 Hash Join  (cost=112.50..215.14 rows=1000 width=16) (actual time=6.635..7.113 rows=1000 loops=1)
   Hash Cond: (a.id = b.id)
   ->  Seq Scan on t_1000 a  (cost=0.00..100.00 rows=1000 width=8) (actual time=2.318..2.554 rows=1000 loops
=1)
   ->  Hash  (cost=100.00..100.00 rows=1000 width=8) (actual time=4.297..4.297 rows=1000 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 48kB
         ->  Seq Scan on t_1000 b  (cost=0.00..100.00 rows=1000 width=8) (actual time=2.142..4.163 rows=1000
 loops=1)
 Planning Time: 0.118 ms
 Execution Time: 7.222 ms
(8 rows)
```
I compared the results with PG, and despite having intermediate costs that are much lower than our intermediate costs, PG has a much higher final total cost. 
```
Nested Loop (cost=0.28..347.44 rows=1000 width=16) (actual time=0.013..1.076 rows=1000 loops=1)
-> Seq Scan on t_1000 a (cost=0.00..15.00 rows=1000 width=8) (actual time=0.006..0.071 rows=1000 loops=1)
-> Index Scan using t_1000_pkey on t_1000 b (cost=0.28..0.33 rows=1 width=8) (actual time=0.001..0.001 rows=1 loops=1000)
Index Cond: (id = a.id)
Planning Time: 0.174 ms
Execution Time: 1.147 ms
```
So I think looking at how all these costs add up to the total cost is probably the key for ensuring the planner doesn't pick the nest loop join.
